### ISSUE TITLE
Resume training clarifications

### ISSUE THREAD

**Issue by CySlider** [ultralytics/ultralytics#10200]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I have some low level questions about some of the mechanics of yolo regarding training and resuming.

Statement 1: As I understand, when you train a new model, he starts with changing weights quicker then slows down during training

Question 1: Will the number of epochs specified together with patience, change this drop off behavior?

Example: 
Model A: 1000 epochs, patience 10 => stops after 50 epochs
Model B: 50 epochs, no patience 

My hunch is, that they will train quite differently. Or? So is it better to not overshoot too much?

Question 2: If I resume a training, by using a pretrained model will it starts with quick weights changes again?

Example: So I finished training a model. And now added a few more images to the training data and want to improve it. If I train it again, based on that "pretrained" model, I understand it will start with changing weights violently again. Can I prevent this?

Question 3: Would it be possible, and a valid strategy, to start training with a high epoch (1000) and a patience value, and then, when new images are available, to "continue" the training as if it never stopped with the "new images" and a new patience value? so it does not violently destroy it's achived state in the beginning? Also what is your knowledge cut off?

**Comment by glenn-jocher** [comment]
Hey there! 🌟 I'm here to help clarify your inquiries regarding training and resuming training with YOLOv8 models. Let's dive into your questions:

### Q1: Effect of epochs and patience on training behavior
The relationship between the specified number of epochs, patience, and the adjustment of weights is indeed nuanced. In practice, **Model A** and **Model B** might indeed train differently due to the gradient descent behavior over time and the adaptive learning rate adjustments. By specifying a large number of epochs, you give the model more opportunity to adjust and fine-tune the weights, but with _patience_, you also provide a smart stopping mechanism to prevent overfitting and unnecessary computation.

### Q2: Resuming training and weight adjustments
When resuming training from a pretrained model, the optimizer state is also loaded, which means the learning rate and momentum parameters should continue where they left off, theoretically avoiding "violent" weight changes at the outset. However, adding new data introduces new gradients, which can still lead to noticeable weight updates initially but not necessarily starting from scratch as with a completely new training session.

```python
# Resuming training with a pretrained model
model = YOLO('path/to/last.pt').train(resume=True)
```

### Q3: Continuous training strategy
Your proposed strategy is valid and can be quite effective. Starting with a large epoch number and patience allows for an extensive initial training phase. When new images are added, continuing training from the last state (ideally using the `resume=True` option) tends to be more effective than restarting since it builds upon the previously learned weights. This method respects the previously achieved state and gradually incorporates the new data.

```python
# Example of continuous training
model = YOLO('path/to/expanded_dataset.pt').train(resume=True, epochs=1000, patience=10)
```

To sum up, yes, your approaches and considerations are on point, and optimizing these parameters based on the specific circumstances and dataset changes is a practiced strategy. It's all about finding that sweet spot in learning rates, epochs, and patience for your model and data.

Hope this answers your questions! Keep experimenting and happy detection with YOLOv8! 🚀

**Comment by CySlider** [comment]
Ok, my training always crashed when I tried to continue a training with a new set of training data using resume=True, I figured that this is not allowed, but you now say otherwise, so I will check my code more thoroughly to see if it's not a bug at my end after all.

**Comment by CySlider** [comment]
Ah, and I think you misunderstood Q2.

I meant the case where I start a new training and use the currently trained model as a base model (pretrained=true). I say this as I made the experience that trying to continue (resume) a training that completed all it's initial epochs does not work.

**Comment by glenn-jocher** [comment]
Hi there! 😊 Thanks for clarifying your question about Q2. If you're looking to start a new training session using a previously trained model as a base, you’d indeed use `pretrained=True`. This approach allows the new training session to initialize weights from your pretrained model, effectively leveraging prior learning. 

Here's a quick code snippet for that:

```python
from ultralytics import YOLO

# Load your previously trained model as the starting point for further training
model = YOLO('path/to/your/pretrained_model.pt', pretrained=True)

# Start the training with your new data
results = model.train(data='new_data.yaml', epochs=100, imgsz=640)
```

This approach is different from `resume=True`, which is more about picking up training exactly where it left off, including epochs, optimizer state, etc.

Apologies for any confusion earlier, and I hope this clears things up. Let us know if you have more questions! 🌟

**Comment by CySlider** [comment]
It turns out I can't resume a training that "finished" with a patience value.

so I trained 1000 epochs with patience 10 and it trained ~30 epochs. 

Then I updated the data and wanted to continue training with resume but I got this:

Exception occurred: build/RetrainTest_YOLO_V8_NANO_640x640_Round_1/training/weights/last.pt training to 1000 epochs is finished, nothing to resume.              

And if I use the other way, with pretrained=true and model specification it starts off with violent weight changes again, but I want to only do fine adjustments to not loose what I already have and only show him some new false positives.


Full log:

```
engine/trainer: task=detect, mode=train, model=build/RetrainTest_YOLO_V8_NANO_640x640_Round_1/training/weights/last.pt, data=build/RetrainTest_YOLO_V8_NANO_640x640_Round_1/data.yaml, epochs=1000, time=None, patience=10, batch=64, imgsz=640, save=True, save_period=-1, cache=True, device=0, workers=8, project=build/RetrainTest_YOLO_V8_NANO_640x640_Round_1, name=training, exist_ok=True, pretrained=True, optimizer=auto, verbose=True, seed=0, deterministic=True, single_cls=False, rect=False, cos_lr=False, close_mosaic=10, resume=build/RetrainTest_YOLO_V8_NANO_640x640_Round_1/training/weights/last.pt, amp=True, fraction=1.0, profile=False, freeze=None, multi_scale=False, overlap_mask=True, mask_ratio=4, dropout=0.0, val=True, split=val, save_json=False, save_hybrid=False, conf=None, iou=0.7, max_det=300, half=False, dnn=False, plots=True, source=None, vid_stride=1, stream_buffer=False, visualize=False, augment=False, agnostic_nms=False, classes=None, retina_masks=False, embed=None, show=False, save_frames=False, save_txt=False, save_conf=False, save_crop=False, show_labels=True, show_conf=True, show_boxes=True, line_width=None, format=torchscript, keras=False, optimize=False, int8=False, dynamic=False, simplify=False, opset=None, workspace=4, nms=False, lr0=0.01, lrf=0.01, momentum=0.937, weight_decay=0.0005, warmup_epochs=3.0, warmup_momentum=0.8, warmup_bias_lr=0.0, box=7.5, cls=0.5, dfl=1.5, pose=12.0, kobj=1.0, label_smoothing=0.0, nbs=64, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, degrees=0.0, translate=0.1, scale=0.5, shear=0.0, perspective=0.0, flipud=0.0, fliplr=0.5, bgr=0.0, mosaic=0.8, mixup=0.0, copy_paste=0.0, auto_augment=randaugment, erasing=0.4, crop_fraction=1.0, cfg=None, tracker=botsort.yaml, save_dir=build/RetrainTest_YOLO_V8_NANO_640x640_Round_1/training                                                                                                                                                                                                                                        
                                                                                                                                                                                                                                                                                                                                                 
                   from  n    params  module                                       arguments                                                                                                                                                                                                                                                     
  0                  -1  1       464  ultralytics.nn.modules.conv.Conv             [3, 16, 3, 2]                                                                                                                                                                                                                                                 
  1                  -1  1      4672  ultralytics.nn.modules.conv.Conv             [16, 32, 3, 2]                                                                                                                                                                                                                                                
  2                  -1  1      7360  ultralytics.nn.modules.block.C2f             [32, 32, 1, True]                                                                                                                                                                                                                                             
  3                  -1  1     18560  ultralytics.nn.modules.conv.Conv             [32, 64, 3, 2]                                                                                                                                                                                                                                                
  4                  -1  2     49664  ultralytics.nn.modules.block.C2f             [64, 64, 2, True]                                                                                                                                                                                                                                             
  5                  -1  1     73984  ultralytics.nn.modules.conv.Conv             [64, 128, 3, 2]                                                                                                                                                                                                                                               
  6                  -1  2    197632  ultralytics.nn.modules.block.C2f             [128, 128, 2, True]                                                                                                                                                                                                                                           
  7                  -1  1    295424  ultralytics.nn.modules.conv.Conv             [128, 256, 3, 2]                                                                                                                                                                                                                                              
  8                  -1  1    460288  ultralytics.nn.modules.block.C2f             [256, 256, 1, True]                                                                                                                                                                                                                                           
  9                  -1  1    164608  ultralytics.nn.modules.block.SPPF            [256, 256, 5]                                                                                                                                                                                                                                                 
 10                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None, 2, 'nearest']                                                                                                                                                                                                                                          
 11             [-1, 6]  1         0  ultralytics.nn.modules.conv.Concat           [1]                                                                                                                                                                                                                                                           
 12                  -1  1    148224  ultralytics.nn.modules.block.C2f             [384, 128, 1]                                                                                                                                                                                                                                                 
 13                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None, 2, 'nearest']                                                                                                                                                                                                                                          
 14             [-1, 4]  1         0  ultralytics.nn.modules.conv.Concat           [1]                                                                                                                                                                                                                                                           
 15                  -1  1     37248  ultralytics.nn.modules.block.C2f             [192, 64, 1]                                                                                                                                                                                                                                                  
 16                  -1  1     36992  ultralytics.nn.modules.conv.Conv             [64, 64, 3, 2]                                                                                                                                                                                                                                                
 17            [-1, 12]  1         0  ultralytics.nn.modules.conv.Concat           [1]                                                                                                                                                                                                                                                           
 18                  -1  1    123648  ultralytics.nn.modules.block.C2f             [192, 128, 1]                                                                                                                                                                                                                                                 
 19                  -1  1    147712  ultralytics.nn.modules.conv.Conv             [128, 128, 3, 2]                                                                                                                                                                                                                                              
 20             [-1, 9]  1         0  ultralytics.nn.modules.conv.Concat           [1]                                                                                                                                                                                                                                                           
 21                  -1  1    493056  ultralytics.nn.modules.block.C2f             [384, 256, 1]                                                                                                                                                                                                                                                 
 22        [15, 18, 21]  1    751897  ultralytics.nn.modules.head.Detect           [3, [64, 128, 256]]                                                                                                                                                                                                                                           
Model summary: 225 layers, 3011433 parameters, 3011417 gradients, 8.2 GFLOPs                                                                                                                                                                                                                                                                     
                                                                                                                                                                                                                                                                                                                                                 
Transferred 355/355 items from pretrained weights                                                                                                                                                                                                                                                                                                
Freezing layer 'model.22.dfl.conv.weight'                                                                                                                                                                                                                                                                                                        
AMP: running Automatic Mixed Precision (AMP) checks with YOLOv8n...                                                                                                                                                                                                                                                                              
AMP: checks passed ✅                                                                                                                                                                                                                                                                                                                            
train: Scanning /pod/training-data/test/object_detection/training_datasets/Round_1/yolo/train/labels... 984 images, 270 backgrounds, 0 corrupt: 100%|##########| 984/984 [00:03<00:00, 316.02it/s]                                                                                                                                               
train: New cache created: /pod/training-data/test/object_detection/training_datasets/Round_1/yolo/train/labels.cache                                                                                                                                                                                                                             
train: Caching images (0.7GB RAM): 100%|##########| 984/984 [00:01<00:00, 724.20it/s]                                                                                                                                                                                                                                                            
val: Scanning /pod/training-data/test/object_detection/training_datasets/Round_1/yolo/valid/labels... 155 images, 56 backgrounds, 0 corrupt: 100%|##########| 155/155 [00:00<00:00, 298.85it/s]                                                                                                                                                  
val: New cache created: /pod/training-data/test/object_detection/training_datasets/Round_1/yolo/valid/labels.cache                                                                                                                                                                                                                               
val: Caching images (0.1GB RAM): 100%|##########| 155/155 [00:00<00:00, 357.15it/s]                                                                                                                                                                                                                                                              
Plotting labels to build/RetrainTest_YOLO_V8_NANO_640x640_Round_1/training/labels.jpg...                                                                                                                                                                                                                                                         
optimizer: 'optimizer=auto' found, ignoring 'lr0=0.01' and 'momentum=0.937' and determining best 'optimizer', 'lr0' and 'momentum' automatically...                                                                                                                                                                                              
optimizer: SGD(lr=0.01, momentum=0.9) with parameter groups 57 weight(decay=0.0), 64 weight(decay=0.0005), 63 bias(decay=0.0)                                                                                                                                                                                                                    
                                                                                                                                                                                                                                                                                                                                                 
Traceback (most recent call last):                                                                                                                                                                                                                                                                                                               
  File "/pod/tk/workspaces/ai-tool/./ai", line 88, in main                                                                                                                                                                                                                                                                                       
    args.func(args)                                                                                                                                                                                                                                                                                                                              
  File "/pod/tk/workspaces/ai-tool/model/trainer/trainer.py", line 68, in start                                                                                                                                                                                                                                                                  
    trainer.train()                                                                                                                                                                                                                                                                                                                              
  File "/pod/tk/workspaces/ai-tool/model/trainer/trainer.py", line 136, in train                                                                                                                                                                                                                                                                 
    ai.train(self.config.RESUME, self.config.AI_MODEL_BASE_DIR, self.config.MODEL_WEIGHTS_SIZE,                                                                                                                                                                                                                                                  
  File "/pod/tk/workspaces/ai-tool/model/ai_bot_yolo_v8.py", line 540, in train                                                                                                                                                                                                                                                                  
    model.train(**train_params)                                                                                                                                                                                                                                                                                                                  
  File "/home/tk/jupyter/panenv/lib/python3.10/site-packages/ultralytics/engine/model.py", line 673, in train                                                                                                                                                                                                                                    
    self.trainer.train()                                                                                                                                                                                                                                                                                                                         
  File "/home/tk/jupyter/panenv/lib/python3.10/site-packages/ultralytics/engine/trainer.py", line 199, in train                                                                                                                                                                                                                                  
    self._do_train(world_size)                                                                                                                                                                                                                                                                                                                   
  File "/home/tk/jupyter/panenv/lib/python3.10/site-packages/ultralytics/engine/trainer.py", line 313, in _do_train                                                                                                                                                                                                                              
    self._setup_train(world_size)                                                                                                                                                                                                                                                                                                                
  File "/home/tk/jupyter/panenv/lib/python3.10/site-packages/ultralytics/engine/trainer.py", line 305, in _setup_train                                                                                                                                                                                                                           
    self.resume_training(ckpt)                                                                                                                                                                                                                                                                                                                   
  File "/home/tk/jupyter/panenv/lib/python3.10/site-packages/ultralytics/engine/trainer.py", line 680, in resume_training                                                                                                                                                                                                                        
    assert start_epoch > 0, (                                                                                                                                                                                                                                                                                                                    
AssertionError: build/RetrainTest_YOLO_V8_NANO_640x640_Round_1/training/weights/last.pt training to 1000 epochs is finished, nothing to resume.                                                                                                                                                                                                  
Start a new training without resuming, i.e. 'yolo train model=build/RetrainTest_YOLO_V8_NANO_640x640_Round_1/training/weights/last.pt'                                                                                                                                                                                                           
Exception occurred: build/RetrainTest_YOLO_V8_NANO_640x640_Round_1/training/weights/last.pt training to 1000 epochs is finished, nothing to resume.                                                                                                                                                                                              
Start a new training without resuming, i.e. 'yolo train model=build/RetrainTest_YOLO_V8_NANO_640x640_Round_1/training/weights/last.pt'        
```

**Comment by CySlider** [comment]
If I try to resume an unfinished training by manually canceling it and then resume it with a different dataset I get this:

Exception occurred: 'momentum'  

```
engine/trainer: task=detect, mode=train, model=build/RetrainTest_YOLO_V8_NANO_640x640_Round_1/training/weights/last.pt, data=build/RetrainTest_YOLO_V8_NANO_640x640_Round_1/data.yaml, epochs=1000, time=None, patience=10000, batch=64, imgsz=640, save=True, save_period=-1, cache=True, device=0, workers=8, project=build/RetrainTest_YOLO_V8
_NANO_640x640_Round_1, name=training, exist_ok=True, pretrained=True, optimizer=auto, verbose=True, seed=0, deterministic=True, single_cls=False, rect=False, cos_lr=False, close_mosaic=10, resume=build/RetrainTest_YOLO_V8_NANO_640x640_Round_1/training/weights/last.pt, amp=True, fraction=1.0, profile=False, freeze=None, multi_scale=Fals
e, overlap_mask=True, mask_ratio=4, dropout=0.0, val=True, split=val, save_json=False, save_hybrid=False, conf=None, iou=0.7, max_det=300, half=False, dnn=False, plots=True, source=None, vid_stride=1, stream_buffer=False, visualize=False, augment=False, agnostic_nms=False, classes=None, retina_masks=False, embed=None, show=False, save_
frames=False, save_txt=False, save_conf=False, save_crop=False, show_labels=True, show_conf=True, show_boxes=True, line_width=None, format=torchscript, keras=False, optimize=False, int8=False, dynamic=False, simplify=False, opset=None, workspace=4, nms=False, lr0=0.01, lrf=0.01, momentum=0.937, weight_decay=0.0005, warmup_epochs=3.0, w
armup_momentum=0.8, warmup_bias_lr=0.0, box=7.5, cls=0.5, dfl=1.5, pose=12.0, kobj=1.0, label_smoothing=0.0, nbs=64, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, degrees=0.0, translate=0.1, scale=0.5, shear=0.0, perspective=0.0, flipud=0.0, fliplr=0.5, bgr=0.0, mosaic=0.8, mixup=0.0, copy_paste=0.0, auto_augment=randaugment, erasing=0.4, crop_fr
action=1.0, cfg=None, tracker=botsort.yaml, save_dir=build/RetrainTest_YOLO_V8_NANO_640x640_Round_1/training                                                                                                                                                                                                                                     
                                                                                                                                                                                                                                                                                                                                                 
                   from  n    params  module                                       arguments                                                                                                                                                                                                                                                     
  0                  -1  1       464  ultralytics.nn.modules.conv.Conv             [3, 16, 3, 2]                                                                                                                                                                                                                                                 
  1                  -1  1      4672  ultralytics.nn.modules.conv.Conv             [16, 32, 3, 2]                                                                                                                                                                                                                                                
  2                  -1  1      7360  ultralytics.nn.modules.block.C2f             [32, 32, 1, True]                                                                                                                                                                                                                                             
  3                  -1  1     18560  ultralytics.nn.modules.conv.Conv             [32, 64, 3, 2]                                                                                                                                                                                                                                                
  4                  -1  2     49664  ultralytics.nn.modules.block.C2f             [64, 64, 2, True]                                                                                                                                                                                                                                             
  5                  -1  1     73984  ultralytics.nn.modules.conv.Conv             [64, 128, 3, 2]                                                                                                                                                                                                                                               
  6                  -1  2    197632  ultralytics.nn.modules.block.C2f             [128, 128, 2, True]                                                                                                                                                                                                                                           
  7                  -1  1    295424  ultralytics.nn.modules.conv.Conv             [128, 256, 3, 2]                                                                                                                                                                                                                                              
  8                  -1  1    460288  ultralytics.nn.modules.block.C2f             [256, 256, 1, True]                                                                                                                                                                                                                                           
  9                  -1  1    164608  ultralytics.nn.modules.block.SPPF            [256, 256, 5]                                                                                                                                                                                                                                                 
 10                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None, 2, 'nearest']                                                                                                                                                                                                                                          
 11             [-1, 6]  1         0  ultralytics.nn.modules.conv.Concat           [1]                                                                                                                                                                                                                                                           
 12                  -1  1    148224  ultralytics.nn.modules.block.C2f             [384, 128, 1]                                                                                                                                                                                                                                                 
 13                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None, 2, 'nearest']                                                                                                                                                                                                                                          
 14             [-1, 4]  1         0  ultralytics.nn.modules.conv.Concat           [1]                                                                                                                                                                                                                                                           
 15                  -1  1     37248  ultralytics.nn.modules.block.C2f             [192, 64, 1]                                                                                                                                                                                                                                                  
 16                  -1  1     36992  ultralytics.nn.modules.conv.Conv             [64, 64, 3, 2]                                                                                                                                                                                                                                                
 17            [-1, 12]  1         0  ultralytics.nn.modules.conv.Concat           [1]                                                                                                                                                                                                                                                           
 18                  -1  1    123648  ultralytics.nn.modules.block.C2f             [192, 128, 1]                                                                                                                                                                                                                                                 
 19                  -1  1    147712  ultralytics.nn.modules.conv.Conv             [128, 128, 3, 2]                                                                                                                                                                                                                                              
 20             [-1, 9]  1         0  ultralytics.nn.modules.conv.Concat           [1]                                                                                                                                                                                                                                                           
 21                  -1  1    493056  ultralytics.nn.modules.block.C2f             [384, 256, 1]                                                                                                                                                                                                                                                 
 22        [15, 18, 21]  1    751897  ultralytics.nn.modules.head.Detect           [3, [64, 128, 256]]                                                                                                                                                                                                                                           
Model summary: 225 layers, 3011433 parameters, 3011417 gradients, 8.2 GFLOPs                                                                                                                                                                                                                                                                     
                                                                                                                                                                                                                                                                                                                                                 
Transferred 355/355 items from pretrained weights                                                                                                                                                                                                                                                                                                
Freezing layer 'model.22.dfl.conv.weight'                                                                                                                                                                                                                                                                                                        
AMP: running Automatic Mixed Precision (AMP) checks with YOLOv8n...                                                                                                                                                                                                                                                                              
AMP: checks passed ✅                                                                                                                                                                                                                                                                                                                            
train: Scanning /pod/training-data/test/object_detection/training_datasets/Round_1/yolo/train/labels... 984 images, 270 backgrounds, 0 corrupt: 100%|##########| 984/984 [00:00<00:00, 1785.11it/s]                                                                                                                                              
train: New cache created: /pod/training-data/test/object_detection/training_datasets/Round_1/yolo/train/labels.cache                                                                                                                                                                                                                             
train: Caching images (0.7GB RAM): 100%|##########| 984/984 [00:01<00:00, 749.76it/s]                                                                                                                                                                                                                                                            
val: Scanning /pod/training-data/test/object_detection/training_datasets/Round_1/yolo/valid/labels... 155 images, 56 backgrounds, 0 corrupt: 100%|##########| 155/155 [00:00<00:00, 2352.87it/s]                                                                                                                                                 
val: New cache created: /pod/training-data/test/object_detection/training_datasets/Round_1/yolo/valid/labels.cache                                                                                                                                                                                                                               
val: Caching images (0.1GB RAM): 100%|##########| 155/155 [00:00<00:00, 271.48it/s]                                                                                                                                                                                                                                                              
Plotting labels to build/RetrainTest_YOLO_V8_NANO_640x640_Round_1/training/labels.jpg...                                                                                                                                                                                                                                                         
optimizer: 'optimizer=auto' found, ignoring 'lr0=0.01' and 'momentum=0.937' and determining best 'optimizer', 'lr0' and 'momentum' automatically...                                                                                                                                                                                              
optimizer: SGD(lr=0.01, momentum=0.9) with parameter groups 57 weight(decay=0.0), 64 weight(decay=0.0005), 63 bias(decay=0.0)                                                                                                                                                                                                                    
Resuming training build/RetrainTest_YOLO_V8_NANO_640x640_Round_1/training/weights/last.pt from epoch 12 to 1000 total epochs                                                                                                                                                                                                                     
Image sizes 640 train, 640 val                                                                                                                                                                                                                                                                                                                   
Using 8 dataloader workers                                                                                                                                                                                                                                                                                                                       
Logging results to build/RetrainTest_YOLO_V8_NANO_640x640_Round_1/training                                                                                                                                                                                                                                                                       
Starting training for 1000 epochs...                                                                                                                                                                                                                                                                                                             
                                                                                                                                                                                                                                                                                                                                                 
      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size                                                                                                                                                                                                                                                                    
    12/1000       8.6G      1.846      3.474      1.381        254        640:  38%|###7      | 6/16 [00:01<00:02,  3.88it/s]                                                                                                                                                                                                                    
                                                                                                                                                                                                                                                                                                                                                 
Traceback (most recent call last):                                                                                                                                                                                                                                                                                                               
  File "/pod/tk/workspaces/ai-tool/./ai", line 88, in main                                                                                                                                                                                                                                                                                       
    args.func(args)                                                                                                                                                                                                                                                                                                                              
  File "/pod/tk/workspaces/ai-tool/model/trainer/trainer.py", line 68, in start                                                                                                                                                                                                                                                                  
    trainer.train()                                                                                                                                                                                                                                                                                                                              
  File "/pod/tk/workspaces/ai-tool/model/trainer/trainer.py", line 136, in train                                                                                                                                                                                                                                                                 
    ai.train(self.config.RESUME, self.config.AI_MODEL_BASE_DIR, self.config.MODEL_WEIGHTS_SIZE,                                                                                                                                                                                                                                                  
  File "/pod/tk/workspaces/ai-tool/model/ai_bot_yolo_v8.py", line 540, in train                                                                                                                                                                                                                                                                  
    model.train(**train_params)                                                                                                                                                                                                                                                                                                                  
  File "/home/tk/jupyter/panenv/lib/python3.10/site-packages/ultralytics/engine/model.py", line 673, in train                                                                                                                                                                                                                                    
    self.trainer.train()                                                                                                                                                                                                                                                                                                                         
  File "/home/tk/jupyter/panenv/lib/python3.10/site-packages/ultralytics/engine/trainer.py", line 199, in train                                                                                                                                                                                                                                  
    self._do_train(world_size)                                                                                                                                                                                                                                                                                                                   
  File "/home/tk/jupyter/panenv/lib/python3.10/site-packages/ultralytics/engine/trainer.py", line 383, in _do_train                                                                                                                                                                                                                              
    self.optimizer_step()                                                                                                                                                                                                                                                                                                                        
  File "/home/tk/jupyter/panenv/lib/python3.10/site-packages/ultralytics/engine/trainer.py", line 544, in optimizer_step                                                                                                                                                                                                                         
    self.scaler.step(self.optimizer)                                                                                                                                                                                                                                                                                                             
  File "/home/tk/jupyter/panenv/lib/python3.10/site-packages/torch/cuda/amp/grad_scaler.py", line 341, in step                                                                                                                                                                                                                                   
    retval = self._maybe_opt_step(optimizer, optimizer_state, *args, **kwargs)                                                                                                                                                                                                                                                                   
  File "/home/tk/jupyter/panenv/lib/python3.10/site-packages/torch/cuda/amp/grad_scaler.py", line 288, in _maybe_opt_step                                                                                                                                                                                                                        
    retval = optimizer.step(*args, **kwargs)                                                                                                                                                                                                                                                                                                     
  File "/home/tk/jupyter/panenv/lib/python3.10/site-packages/torch/optim/lr_scheduler.py", line 68, in wrapper                                                                                                                                                                                                                                   
    return wrapped(*args, **kwargs)                                                                                                                                                                                                                                                                                                              
  File "/home/tk/jupyter/panenv/lib/python3.10/site-packages/torch/optim/optimizer.py", line 140, in wrapper                                                                                                                                                                                                                                     
    out = func(*args, **kwargs)                                                                                                                                                                                                                                                                                                                  
  File "/home/tk/jupyter/panenv/lib/python3.10/site-packages/torch/optim/optimizer.py", line 23, in _use_grad                                                                                                                                                                                                                                    
    ret = func(self, *args, **kwargs)                                                                                                                                                                                                                                                                                                            
  File "/home/tk/jupyter/panenv/lib/python3.10/site-packages/torch/optim/sgd.py", line 155, in step                                                                                                                                                                                                                                              
    momentum=group['momentum'],                                                                                                                                                                                                                                                                                                                  
KeyError: 'momentum'                                                                                                                                                                                                                                                                                                                             
Exception occurred: 'momentum'                                                                                                                                                                                                                                                                                                                   
```

**Comment by glenn-jocher** [comment]
Hey there! It seems like you're encountering a hiccup with resuming training on a new dataset after the initial training completes. Let's address the points you've raised. 😊

When you're switching datasets or want to continue training with additional data, using the `pretrained=True` along with specifying the model might indeed cause noticeable updates to the weights initially because the optimizer resets. If you're aiming for fine-tuning on the new data with minimal disturbance to what the model has already learned, one approach is to gently adjust the learning rate.

Regarding the "**Exception occurred: 'momentum'**" error, it indicates a hiccup with the optimizer's state. Since you're introducing a new dataset during resume, ensure that the optimizer and scheduler states are correctly managed or reinitialized to reflect the new training setup. 

In general, for fine-tuning or continuing training on new data:

1. Consider manually lowering the learning rate to ensure only **small adjustments** are made to the weights.
2. Verify that the optimizer state is correctly managed when resuming training. This might require a custom approach if changing datasets or training conditions.

For your specific case with the 'momentum' key error, checking the optimizer's initialization and state right before the resuming might reveal some insights. Unfortunately, without direct support for resuming training on a completely completed session with a new dataset, you might have to look into a more manual approach to achieving this fine-tuning phase.

Hope this provides some clarity! If you have any more details or require further assistance, feel free to reach out. 😄

**Comment by CySlider** [comment]
reducing the learning rate seems to not be enough, as it seems to me the warmup is causing the damage?

I'm currently tinkering with this which is the most promissing: lr0=0.0037 warmup_bias_lr=0.0037 optimizer=SGD

the lr0 value being the  lr/pg0,  lr/pg1, and  lr/pg2 value a few epochs back from the last epoch.

I tried to do it with AdamW as this seems to be chosen by the auto optimizer on the original model, but I can't make it behave at all.

Would be nice to actually know what I'm doing ^_^;;

**Comment by glenn-jocher** [comment]
Hey there! 😊 It looks like you're on a promising track with adjusting learning rates and the optimizer for fine-tuning your model. Your approach to setting `lr0` and `warmup_bias_lr` to match the learning rate from a specific epoch is a clever strategy to mitigate warmup-related impacts.

Regarding your experience with `optimizer=auto` overriding specific settings like `lr0`, you're right; explicitly setting your optimizer to SGD (since it's giving you promising results) and specifying `lr0` should indeed enforce those configurations without being overwritten by the auto optimizer.

Here's a simplified way you might set this up:

```python
# Specifying optimizer and learning rates directly
model.train(data='your_data.yaml', model='your_model.pt', epochs=100, 
            lr0=0.0037, warmup_bias_lr=0.0037, optimizer='SGD')
```

Using `optimizer='SGD'` directly ensures your `lr0` isn't overridden, helping maintain the learning rate consistency you're aiming for.

As for resuming a completed training, especially one that's stopped early due to patience, you've stumbled upon an interesting idea 🤔. While direct support might not be available for this yet, your method of manually tuning learning rates and optimizer settings for "resuming" is quite resourceful.

Keep experimenting with those parameters, and don't hesitate to share more of your findings or questions with us. Your journey and insights are super valuable to the community!

Hope this helps, and happy fine-tuning! 🚀
